Technote organisational structure (#443)
apollographql/docs: src/content/technotes/TN0041-organisational-structure.mdx

---
title: Organizational Structure to support a Supergraph
id: TN0041
description: How to best organise your engineering teams when delivering a Supergraph
tags: [federation, subgraphs, best-practices, organisation]
---

> When setting out to deliver a federated Supergraph at scale, one of the biggest hurdles organisations face is often not the technical understanding, but rather adopting an appropriate organisational structure to support it. There are a number of ways in which organisations have adapted to best enable the delivery of the Supergraph, which we'll look to discuss here in the context of each of the different stages of delivery. However, it is worth first re-addressing the primary drivers and value of adopting a federated graph to help provide context for the organisational decisions proposed and how they fit into the bigger picture.

Reviewer: Im going to be pedantic. Do you need to "sell" values in a tech note, or keep it as focused to the solution.

Author: Yeah I'm with you, my thinking was that this one was quite context dependant is all - do you think maybe just dive in? Wasn't sure how else to make it flow naturally?

**Schema design and Development**

> The first step in the implementation of a Supergraph is well considered Schema design and Subgraph development. In this phase the primary concern when building out a Schema is the needs and requirements of the client. This requires the engineers responsible for their contributions to the Schema to truly understand the client requirements and be able to provide an efficient implementation for retrieving and providing the given data. Often this can lend itself to **teams that work vertically within given domains/features to allow collaboration** between FE Engineers, User Researchers, Designers as well as autonomy and ownership of their Subgraph that supports the given domain. It’s worth noting that there are of course situations where a GraphQL endpoint is being developed prior to clients needing to consume it, however in these situations (i.e. public facing GraphQL endpoints for future public consumption) the way in which data is likely to be consumed can still be considered. For instance, it is important that Schemas are human readable to allow clients to understand the data available, as well as making sure Queries and Mutations are specific and that fields are clear and well documented to ensure that clients are able to get maximum value from the graph. Ultimately, when organising engineering teams to build out and implement each of the Subgraphs that make up the overall Supergraph, the goal is well governed autonomous teams.

Reviewer 2: ```suggestion
The first step in the implementation of a supergraph is well considered schema design and subgraph development. In this phase the primary concern when building out a schema is the needs and requirements of the client. This requires the engineers responsible for their contributions to the schema to truly understand the client requirements and be able to provide an efficient implementation for retrieving and providing the given data. 

Often this can lend itself to **teams that work vertically within given domains/features to allow collaboration** between FE Engineers, User Researchers, Designers as well as autonomy and ownership of their subgraph that supports the given domain. It’s worth noting that there are of course situations where a GraphQL endpoint is being developed prior to clients needing to consume it, however in these situations (i.e. public facing GraphQL endpoints for future public consumption) the way in which data is likely to be consumed can still be considered. For instance, it is important that schemas are human readable to allow clients to understand the data available, as well as making sure queries and mutations are specific and that fields are clear and well documented to ensure that clients are able to get maximum value from the graph. Ultimately, when organising engineering teams to build out and implement each of the subgraphs that make up the overall supergraph, the goal is well governed autonomous teams.
```

> The first step in the implementation of a supergraph is well considered schema design and subgraph development. 

Is confusing, unsure what you're trying to convey here. 

Also as a general note, would recommend checking https://apollographql.quip.com/uSoRAk1r4knK/Apollo-Content-Style-Guide for capitalizations.

Author: Agree, think I was probably listing out the sections with this just being the first one.. but agree it's possibly miss-leading, have updated :)

**Graph Infrastructure**

> The natural next area for consideration is the infrastructure supporting the graph. It is important that there is conscious thought and decision making regarding the graph's infrastructure, either via a "platform team" or at least a collective **governing how the graph infrastructure is configured** (i.e. router deployment, CI integration, developer experience etc..) that has a good view of the components of the Supergraph that are required. One crucial element is making sure that versions are well managed, for example ensuring that the correct version of the federation spec is being used and aligns with consistent versioning of the underlying frameworks being leveraged across Subgraphs, such as Apollo Server to enable and maintain specific features for subgraph teams.

Reviewer 2: > either via a "platform team" or at least a collective **governing

Feels like a missing word here.

However, in addition to promoting GraphQL and federation within engineering teams, one of the biggest hurdles and risks to running GraphQL as part of your stack is **ensuring that the wider business, stakeholders and architects understand the value and the benefits of GraphQL**. Having a solid "Elevator pitch" such as "_It provides a reliable, robust abstraction layer between Front and Backend that enables governance and consistency that was previously unachievable with disparate microservices, allowing reuse, optimisation and cost savings_." can help secure the use of the graph and possibly even help extend the coverage of the Supergraph - and with that, its overall value to the business.

**Conclusion**

Ultimately organisational structure is critical to the success of the graph within an enterprise. Below are some of the key takeaways for setting up your teams for success:

> - Important that teams are setup to allow autonomy over updates to their Subgraph(s).

Reviewer 2: ```suggestion
- It is important that teams are setup to allow autonomy over updates to their subgraph(s).
```

> - CICD should be setup to ensure comprehensive checks and quick feedback to engineers on changes they are trying to push through to Supergraph.

Reviewer 2: ```suggestion
- CI/CD should be setup to ensure comprehensive checks and quick feedback to engineers on changes they are trying to push through to the supergraph.
```
